cdc_vaccines/delphi_cdc_vaccines/pull.py

```diff
@@ -0,0 +1,162 @@
+      from December 13th 2020 until the latest date
+
+    Note that the raw dataset gives the `cumulative` metrics, from which
+    we compute `counts` by taking first differences.  Hence, `counts`
+    may be negative.  This is wholly dependent on the quality of the raw
+    dataset.
+
+    We filter the data such that we only keep rows with valid FIPS, or "FIPS"
+    codes defined under the exceptions of the README.  The current  exceptions
+    include:
+    # - 0: statewise unallocated
+    Parameters
+    ----------
+    base_url: str
+        Base URL for pulling the CDC Vaccination Data
+    logger: Logger
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe as described above.
+    """
+    # Columns to drop the the data frame.
+    drop_columns = [
+    "date",
+    "recip_state",
+    "series_complete_pop_pct",
+    "mmwr_week",
+    "recip_county",
+    "state_id"
+    ]
+
+
+    # Read data
+    df = pd.read_csv(base_url)
+    logger.info("data retrieved from source",
+                num_rows=df.shape[0],
+                num_cols=df.shape[1],
+                min_date=min(df['Date']),
+                max_date=max(df['Date']),
+                checksum=hashlib.sha256(pd.util.hash_pandas_object(df).values).hexdigest())
+    df.columns = [i.lower() for i in df.columns]
+
+    df['recip_state'] = df['recip_state'].str.lower()
+    drop_columns.extend([x for x in df.columns if ("pct" in x) | ("svi" in x)])
+    drop_columns =  list(set(drop_columns))
+    df = GeoMapper().add_geocode(df, "state_id", "state_code",
+        from_col="recip_state", new_col="state_id", dropna=False)
+    df['state_id'] = df['state_id'].fillna('0').astype(int)
+    # Change FIPS from 0 to XX000 for statewise unallocated cases/deaths
+    unassigned_index = (df["fips"] == "UNK")
+    df.loc[unassigned_index, "fips"] = df["state_id"].loc[unassigned_index].values * 1000
+
+    # Conform FIPS
+    df["fips"] = df["fips"].apply(lambda x: f"{int(x):05d}")
+    df["timestamp"] = pd.to_datetime(df["date"])
+    # Drop unnecessary columns (state is pre-encoded in fips)
+    try:
+        df.drop(drop_columns, axis=1, inplace=True)
+    except KeyError as e:
+        raise ValueError(
+            "Tried to drop non-existent columns. The dataset "
+            "schema may have changed.  Please investigate and "
+            "amend drop_columns."
+        ) from e
+    # timestamp: str -> datetime
+    df.columns = ["fips",
+                  "cumulative_counts_tot_vaccine",
+                  "cumulative_counts_tot_vaccine_12P",
+                  "cumulative_counts_tot_vaccine_18P",
+                  "cumulative_counts_tot_vaccine_65P",
+                  "cumulative_counts_part_vaccine",
+                  "cumulative_counts_part_vaccine_12P",
+                  "cumulative_counts_part_vaccine_18P",
+                  "cumulative_counts_part_vaccine_65P",
+                  "timestamp"]
```

Comment:
this longhand can stay; all the ways I can think of to abstract it out are worse